pkg/executor/aggregate/agg_hash_partial_worker.go

```diff
@@ -107,7 +107,66 @@ func (w *HashAggPartialWorker) run(ctx sessionctx.Context, waitGroup *sync.WaitG
 	}
 }
 
-func (w *HashAggPartialWorker) updatePartialResult(ctx sessionctx.Context, sc *stmtctx.StatementContext, chk *chunk.Chunk, _ int) (err error) {
+func (w *HashAggPartialWorker) newPartialResults(n int, partialResultNum int) [][]aggfuncs.PartialResult {
+	if n > len(w.partialResultsBuffer) {
+		appendNum := n - len(w.partialResultsBuffer)
+		for i := 0; i < appendNum; i++ {
+			w.partialResultsBuffer = append(w.partialResultsBuffer, make([]aggfuncs.PartialResult, partialResultNum))
+		}
+	}
+	return w.partialResultsBuffer[0:0]
+}
+
+func (w *HashAggPartialWorker) expandPartialResults(partialResult [][]aggfuncs.PartialResult, partialResultSize int) [][]aggfuncs.PartialResult {
+	partialResultLen := len(partialResult)
+	if len(w.partialResultsBuffer) < partialResultLen+1 {
+		w.partialResultsBuffer = append(w.partialResultsBuffer, make([]aggfuncs.PartialResult, partialResultSize))
+	}
+	return w.partialResultsBuffer[:partialResultLen+1]
+}
+
+// If the group key has appeared before, reuse the partial result.
+// If the group key has not appeared before, create empty partial results.
+func (w *HashAggPartialWorker) getPartialResultsOfEachRow(_ *stmtctx.StatementContext, groupKey [][]byte, mapper []AggPartialResultMapper, finalConcurrency int) [][]aggfuncs.PartialResult {
+	cntOfGroupKeys := len(groupKey)
+	allMemDelta := int64(0)
+	partialResultNum := w.getPartialResultSliceLenConsiderByteAlign()
+	partialResultsOfEachRow := w.newPartialResults(cntOfGroupKeys, partialResultNum)
+
+	for i := 0; i < cntOfGroupKeys; i++ {
+		finalWorkerIdx := int(murmur3.Sum32(groupKey[i])) % finalConcurrency
+		tmp, ok := mapper[finalWorkerIdx][string(groupKey[i])]
```

Comment:
```suggestion
		tmp, ok := mapper[finalWorkerIdx][string(hack.String(groupKey[i])]
```

Reply:
I think it's unsafe to use hack.String as groupKey may be modified.

Reply:
It's safe there because this is a read operation

Reply:
> It's safe there because this is a read operation

However, the `groupKey` may be modified in this [function](https://github.com/pingcap/tidb/pull/47428/files#diff-d0947103b1185ee1d6b4d67efff274ec66d0888b12e363f8ce04fe59233e3b66R158)